File: app/src/main/java/com/zwstudio/logicgamesandroid/bridges/domain/BridgesGameState.java
package com.zwstudio.logicgamesandroid.bridges.domain;

import com.zwstudio.logicgamesandroid.logicgames.domain.CellsGameState;
import com.zwstudio.logicgamesandroid.logicgames.domain.Graph;
import com.zwstudio.logicgamesandroid.logicgames.domain.LogicGamesHintState;
import com.zwstudio.logicgamesandroid.logicgames.domain.Node;
import com.zwstudio.logicgamesandroid.logicgames.domain.Position;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static fj.data.Array.arrayArray;
import static fj.data.List.iterableList;
import static fj.function.Integers.add;

/**
 * Created by zwvista on 2016/09/29.
 */

public class BridgesGameState extends CellsGameState<BridgesGame> {
    public BridgesObject[] objArray;

    public BridgesGameState(BridgesGame game) {
        super(game);
        objArray = new BridgesObject[rows() * cols()];
        Arrays.fill(objArray, new BridgesEmptyObject());
    }

    public BridgesObject get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public BridgesObject get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, BridgesObject obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, BridgesObject obj) {
        set(p.row, p.col, obj);
    }

    private void updateIsSolved() {
        isSolved = true;
        Graph g = new Graph();
        Map<Position, Node> pos2Node = new HashMap<>();
        for (Map.Entry<Position, BridgesIslandInfo> entry : game.islandsInfo.entrySet()) {
            Position p = entry.getKey();
            BridgesIslandInfo info = entry.getValue();
            BridgesIslandObject o = (BridgesIslandObject) get(p);
            int n1 = arrayArray(o.bridges).foldLeft(add, 0);
            int n2 = info.bridges;
            o.state = n1 < n2 ? LogicGamesHintState.Normal : n1 == n2 ? LogicGamesHintState.Complete : LogicGamesHintState.Error;
            if (n1 != n2) isSolved = false;
            if (!isSolved) continue;
            Node node = new Node(p.toString());
            pos2Node.put(p, node);
            g.addNode(node);
        }
        if (!isSolved) return;
        for (Map.Entry<Position, BridgesIslandInfo> entry : game.islandsInfo.entrySet()) {
            Position p = entry.getKey();
            BridgesIslandInfo info = entry.getValue();
            for (Position p2 : info.neighbors) {
                if (p2 == null) continue;
                g.connectNode(pos2Node.get(p), pos2Node.get(p2));
            }
        }
        g.setRootNode(iterableList(pos2Node.values()).head());
        List<Node> nodeList = g.bfs();
        int n1 = nodeList.size();
        int n2 = pos2Node.values().size();
        if (n1 != n2) isSolved = false;
    }

    public boolean switchBridges(BridgesGameMove move) {
        Position pFrom = move.pFrom, pTo = move.pTo;
        if (!(pFrom.compareTo(pTo) < 0 && (pFrom.row == pTo.row || pFrom.col == pTo.col))) return false;
        BridgesObject o11 = get(pFrom), o22 = get(pTo);
        if (!(o11 instanceof BridgesIslandObject && o22 instanceof BridgesIslandObject)) return false;
        BridgesIslandObject o1 = (BridgesIslandObject) o11, o2 = (BridgesIslandObject) o22;
        int n1 = pFrom.row == pTo.row ? 1 : 2;
        int n2 = (n1 + 2) % 4;
        Position os = BridgesGame.offset[n1];
        for (Position p = pFrom.add(os); !p.equals(pTo); p.addBy(os))
            switch (o1.bridges[n1]) {
            case 0:
                {
                    BridgesObject o = get(p);
                    if (!(o instanceof BridgesEmptyObject)) return false;
                    set(p, new BridgesBridgeObject());
                }
                break;
            case 1:
                set(p, new BridgesEmptyObject());
                break;
            }
        int n = (o1.bridges[n1] + 1) % 3;
        o1.bridges[n1] = o2.bridges[n2] = n;
        updateIsSolved();
        return true;
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/bridges/BridgesGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.bridges

import com.zwstudio.logicpuzzlesandroid.common.domain.*

class BridgesGameState(game: BridgesGame) : CellsGameState<BridgesGame, BridgesGameMove, BridgesGameState>(game) {
    var objArray = Array<BridgesObject>(rows * cols) { BridgesEmptyObject }

    operator fun get(row: Int, col: Int) = objArray[row * cols + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, obj: BridgesObject) {objArray[row * cols + col] = obj}
    operator fun set(p: Position, obj: BridgesObject) {this[p.row, p.col] = obj}

    init {
        for (p in game.islandsInfo.keys)
            this[p] = BridgesIslandObject()
    }

    fun switchBridges(move: BridgesGameMove): Boolean {
        val pFrom = move.pFrom
        val pTo = move.pTo
        // 4. Bridges can only run horizontally or vertically.
        if (!(pFrom < pTo && (pFrom.row == pTo.row || pFrom.col == pTo.col))) return false
        val o1 = this[pFrom]
        val o2 = this[pTo]
        if (!(o1 is BridgesIslandObject && o2 is BridgesIslandObject)) return false
        val n1 = if (pFrom.row == pTo.row) 1 else 2
        val n2 = (n1 + 2) % 4
        val os = BridgesGame.offset[n1]
        var p = pFrom + os
        while (p != pTo) {
            when (o1.bridges[n1]) {
                0 -> {
                    // 4. Bridges can't cross each other.
                    if (this[p] !is BridgesEmptyObject) return false
                    this[p] = BridgesBridgeObject
                }
                2 -> this[p] = BridgesEmptyObject
            }
            p += os
        }
        // 5. Lastly, you can connect two islands with either one or two Bridges
        // (or none, of course)
        val n = (o1.bridges[n1] + 1) % 3
        o2.bridges[n2] = n
        o1.bridges[n1] = o2.bridges[n2]
        updateIsSolved()
        return true
    }

    /*
        iOS Game: Logic Games/Puzzle Set 7/Bridges

        Summary
        Enough Sudoku, let's build!

        Description
        1. The board represents a Sea with some islands on it.
        2. You must connect all the islands with Bridges, making sure every
           island is connected to each other with a Bridges path.
        3. The number on each island tells you how many Bridges are touching
           that island.
        4. Bridges can only run horizontally or vertically and can't cross
           each other.
        5. Lastly, you can connect two islands with either one or two Bridges
           (or none, of course)
    */
    private fun updateIsSolved() {
        isSolved = true
        val g = Graph()
        val pos2node = mutableMapOf<Position, Node>()
        // 3. The number on each island tells you how many Bridges are touching
        // that island.
        for ((p, info) in game.islandsInfo) {
            val o = this[p] as BridgesIslandObject
            val n1 = o.bridges.sum()
            val n2 = info.bridges
            o.state = if (n1 < n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            if (n1 != n2) isSolved = false
            if (!isSolved) continue
            val node = Node(p.toString())
            pos2node[p] = node
            g.addNode(node)
        }
        if (!isSolved) return
        for ((p, info) in game.islandsInfo) {
            for (p2 in info.neighbors) {
                if (p2 == null) continue
                g.connectNode(pos2node[p]!!, pos2node[p2]!!)
            }
        }
        // 2. You must connect all the islands with Bridges, making sure every
        // island is connected to each other with a Bridges path.
        g.rootNode = pos2node.values.first()
        val nodeList = g.bfs()
        if (nodeList.size != pos2node.size) isSolved = false
    }
}
